### ISSUE TITLE
ChannelOperationsHandler.drain() - In pendingWrite queue, if polled "future" is null, why not poll "msg" as well?

### ISSUE THREAD

**Issue by get-set** [reactor/reactor-netty#313]
https://github.com/reactor/reactor-netty/blob/10ae80311b56957246b441a3a13c4f9e278ee840/src/main/java/reactor/ipc/netty/channel/ChannelOperationsHandler.java#L391-L409

(This is a question)
The code here, first we get `future` via `pendingWrites.poll()`, and then check if the `future` is null, if so, just `break` or `continue`, why not to poll the `Object`(as a pair with the `future`) afterwards? If just `break` or `continue`, next time when we want `future` but get the object instead, I'm confused here, and by the way, when would the polled `future` be null?

**Comment by simonbasle** [comment]
This is an active drain loop pattern, where the `continue` happens because thanks to the WIP atomic, we detected more work was concurrently added. So the first thread to enter that loop "wins" and steals that additional work that was added in parallel. If no more concurrent work is detected (`decrementAndGet` of `0`) then we exit the loop.

There is no second `poll()` if the `Future` is `null`, because we only ever enqueue the `Future` with a companion `Object`. This is a bit hidden, because that offering of two values is atomic and is disguised as a `BiPredicate.test` call: see calls in the form `pendingWriteOffer.test(future, v)`.

The `Future` could be `null` in the case where absolutely all the enqueued work has been processed.
